packages/blocks/src/types/block-attributes.ts

```diff
@@ -0,0 +1,319 @@
+ *
+ * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/#meta-source-deprecated Meta Source} on developer.wordpress.org
+ * @deprecated
+ * @public
+ */
+export interface Meta {
+	source: 'meta';
+	type: 'string';
+	meta: string;
+	default?: string;
+}
+
+/**
+ * Attribute sourced from array of values from markup.
+ *
+ * Entries of the array are determined by the selector argument, where each matched element within
+ * the block will have an entry structured corresponding to the second argument, an object of
+ * attribute sources.
+ *
+ * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/#query-source Query Source} on developer.wordpress.org
+ * @public
+ */
+export interface Query< T > {
+	source: 'query';
+	type: 'array';
+	selector: string;
+	query: {
+		[ k in keyof T ]: BlockAttribute<
+			T[ k ] extends Array< infer U > ? U : T[ k ]
+		>;
+	};
+	default?: any[];
+}
+
+/**
+ * Use text to extract the inner text from markup.
+ *
+ * Note that HTML is returned according to the rules of {@link https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent textContent }.
+ *
+ * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/#text-source Text Source} on developer.wordpress.org
+ * @public
+ */
+export interface Text {
+	source: 'text';
+	type: 'string';
+	selector?: string;
+	default?: string;
+}
+
+/**
+ * @internal
+ */
+export type None =
+	| {
+			source?: never;
+	  } & (
+			| {
+					type: 'array';
+					default?: any[];
+			  }
+			| {
+					type: 'object';
+					default?: object;
+			  }
+			| {
+					type: 'boolean';
+					default?: boolean;
+			  }
+			| {
+					type: 'number';
+					default?: number;
+			  }
+			| {
+					type: 'string';
+					default?: string;
+			  }
+	   );
+
+/**
+ * @internal
+ */
+export type Primitive = 'array' | 'object' | 'boolean' | 'number' | 'string';
+
+/**
+ * Union of block attribute schemas.
+ *
+ * @public
+ */
+export type BlockAttributeSchema< T > = (
+	| Raw
+	| Attribute
+	| Property
+	| Children
+	| Node
+	| Tag
+	| HTML
+	| Meta
+	| Query< T >
+	| Text
+	| None
+ ) & {
+	/**
+	 * TODO Undocumented
+	 *
+	 * Used in [selectors.js](../store/selectors.js) and [utils.js](../api/utils.js)
+	 *
+	 * @internal
+	 */
+	__experimentalRole?: string;
+};
+
+/**
+ * @public
+ */
+export type BlockAttribute< T > = BlockAttributeSchema< T > | Primitive;
+
+/**
+ * Attributes definition providing information about the data stored by a block.
+ *
+ * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/ Attributes}
+ * @public
+ */
+export type BlockAttributes< T extends Record< string, any > = {} > = Record<
+	string,
+	BlockAttribute< T >
+> & {
+	/**
+	 * TODO Undocumented
```

Comment:
To all of those descriptions, we could add that they're legacy and are no longer in use.

Reply:
I'm not particularly sure, we might have to go through them one by one in history. I don't think it's particularly necessary to add the deprecation version to the types.